File: common/src/main/java/com/ssblur/scriptor/item/Chalk.java
package com.ssblur.scriptor.item;

import com.ssblur.scriptor.events.network.server.ChalkNetwork;
import net.minecraft.ChatFormatting;
import net.minecraft.core.component.DataComponents;
import net.minecraft.network.chat.Component;
import net.minecraft.world.InteractionHand;
import net.minecraft.world.InteractionResultHolder;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.TooltipFlag;
import net.minecraft.world.level.Level;

import java.util.List;

public class Chalk extends Item {
  public Chalk(Properties properties) {
    super(properties);
  }

  @Override
  public void appendHoverText(ItemStack itemStack, TooltipContext level, List<Component> list, TooltipFlag tooltipFlag) {
    super.appendHoverText(itemStack, level, list, tooltipFlag);
    if(!hasName(itemStack)) {
      list.add(Component.translatable("extra.scriptor.chalk_unnamed_1").withStyle(ChatFormatting.RED));
      list.add(Component.translatable("extra.scriptor.chalk_unnamed_2"));
    }
  }

  @Override
  public InteractionResultHolder<ItemStack> use(Level level, Player player, InteractionHand interactionHand) {
    var result = super.use(level, player, interactionHand);

    if(level.isClientSide)
      ChalkNetwork.sendChalkMessage();

    return result;
  }

  public static boolean hasName(ItemStack itemStack) {
    var name = itemStack.get(DataComponents.ITEM_NAME);
    return name != null;
  }
}

File: common/src/main/kotlin/com/ssblur/scriptor/item/tools/Chalk.kt
package com.ssblur.scriptor.item.tools

import com.ssblur.scriptor.network.server.ScriptorNetworkC2S
import com.ssblur.scriptor.network.server.ScriptorNetworkC2S.SendChalk
import net.minecraft.ChatFormatting
import net.minecraft.client.Minecraft
import net.minecraft.core.component.DataComponents
import net.minecraft.network.chat.Component
import net.minecraft.world.InteractionHand
import net.minecraft.world.InteractionResultHolder
import net.minecraft.world.entity.player.Player
import net.minecraft.world.item.Item
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.TooltipFlag
import net.minecraft.world.level.Level
import net.minecraft.world.phys.BlockHitResult

open class Chalk(properties: Properties) : Item(properties) {
    override fun appendHoverText(
        itemStack: ItemStack,
        level: TooltipContext,
        list: MutableList<Component>,
        tooltipFlag: TooltipFlag
    ) {
        super.appendHoverText(itemStack, level, list, tooltipFlag)
        if (!hasName(itemStack)) {
            list.add(Component.translatable("extra.scriptor.chalk_unnamed_1").withStyle(ChatFormatting.RED))
            list.add(Component.translatable("extra.scriptor.chalk_unnamed_2"))
        }
    }

    override fun use(
        level: Level,
        player: Player,
        interactionHand: InteractionHand
    ): InteractionResultHolder<ItemStack> {
        if (level.isClientSide) {
            val client = Minecraft.getInstance()
            val hit = client.hitResult
            if (hit is BlockHitResult)
                ScriptorNetworkC2S.sendChalk(SendChalk(hit, false))
        }

        return InteractionResultHolder.success(player.getItemInHand(interactionHand))
    }

    companion object {
        fun hasName(itemStack: ItemStack): Boolean {
            val name = itemStack[DataComponents.CUSTOM_NAME]
            return name != null
        }
    }
}
